### PR TITLE
Steal Array.prototype methods for List and Results

### PR DESCRIPTION
**Pull request by appden**
This pretty much steals all non-mutating methods that were possible to take. The stealing is done in JS so it can be compatible with Chrome debugging as well as the future V8 binding. In order to do this, the `Realm.List` and `Realm.Results` constructors were exposed (meaning you can perform `instanceof` type checks now), which will throw if called directly.

This specifically excludes `indexOf`, `lastIndexOf`, and `includes` because those require will require special implementations in another PR.

### CODE REVIEW

**Comment by alazier** [comment]
As we discussed it would be preferable to not spread the binding implementation across JS and c. I don't see this would not work in chrome debugging if we implemented this using JSC.

**Comment by alazier** [comment]
I can see why we would need to do the same in chrome to make debugging work as passing callbacks over rpc isn't possible. Still not convinced we shouldn't be stealing the methods in jsc when running on the device.

File: lib/browser/results.js [hunk 1]
```diff
@@ -18,21 +18,21 @@
...
+createMethods(Results.prototype, objectTypes.RESULTS, [
     'filtered',
     'sorted',
     'snapshot',
 ]);
 
-function create(realmId, info) {
-    return util.createList(Results.prototype, realmId, info);
+export function create(realmId, info) {
+    return createList(Results.prototype, realmId, info);
```
Comment (alazier) [inline lib/browser/results.js]: maybe this should named differently since it isn't creating a list

**Comment by alazier** [comment]
Would make this a bit easier to review if the js/require changes were a separate pr. Fine for now but keep in mind for next time.

File: src/js_init.cpp [hunk 2]
```diff
@@ -54,27 +56,36 @@ JSClassRef RJSRealmTypeClass() {
     return JSClassCreate(&realmTypesDefinition);
 }
 
+static JSObjectRef InvalidConstructor(JSContextRef ctx, JSObjectRef constructor, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception) {
```
Comment (alazier) [inline src/js_init.cpp]: This name is a bit confusing. Maybe call it UncallableConstructor

File: src/js_util.hpp [hunk 3]
```diff
@@ -201,6 +201,14 @@ static inline size_t RJSValidatedListLength(JSContextRef ctx, JSObjectRef object
     return RJSValidatedValueToNumber(ctx, lengthValue);
 }
 
+static inline void RJSObjectSetProperty(JSContextRef ctx, JSObjectRef object, JSStringRef propertyName, JSValueRef value, JSPropertyAttributes attributes = 0) {
```
Comment (alazier) [inline src/js_util.hpp]: Should include the word Validated for naming consistency

**Comment by alazier** [comment]
So other than my comments this looks good. CI is having issues so we should run locally to confirm everything is working.

**Comment by appden** [comment]
All tests pass locally.

**Pull request merged**